### PR TITLE
Added TrainingHubCallback abstraction and Unsloth adapter

### PR DESCRIPTION
**Pull request by Sridhar1030**
## Summary
- Add a unified `TrainingHubCallback` base class and `TrainingHubContext` dataclass for backend-agnostic training lifecycle hooks
- Add Unsloth/HuggingFace adapter (`UnslothCallbackAdapter`) with exception isolation and rank-0 guard
- Wire `callbacks` and `eval_data_path` into `lora_sft` / Unsloth LoRA backend so user callbacks fire during real training
- Add unit tests and an Unsloth smoke example (script + notebook)

This is the first split of [RHOAIENG-54744](https://redhat.atlassian.net/browse/RHOAIENG-54744). InstructLab / Mini-Trainer adapters and torchrun serialization are deferred to [RHOAIENG-77627](https://redhat.atlassian.net/browse/RHOAIENG-77627).

## What's included
| Piece | Details |
| --- | --- |
| `TrainingHubCallback` | 9 no-op lifecycle hooks (`on_train_begin` … `on_train_end`) |
| `TrainingHubContext` | Normalized `step`, `epoch`, `loss`, `learning_rate`, `is_main_process`, `output_dir`, `metrics` |
| Unsloth adapter | Wraps hub callbacks as HF `TrainerCallback`; forwards eval metrics on `on_evaluate` |
| API | `lora_sft(..., callbacks=[...], eval_data_path=...)` |
| Unit tests | `tests/test_callbacks.py` — context, hooks, adapter dispatch, exception isolation, rank guard, eval metrics |
| Smoke notebook | `examples/notebooks/callback_smoke_unsloth.ipynb` — guided Unsloth e2e cells (setup, train+eval, pass/fail for all 9 hooks) |
| Smoke script | `examples/scripts/callback_smoke_unsloth.py` — same check runnable from a workbench terminal |
## Test plan
- [x] `pytest tests/test_callbacks.py` (15 passed)
- [x] OpenShift A100 smoke with Unsloth + `Qwen2.5-0.5B-Instruct`
  - Confirmed all 9 unified hooks fired: begin/epoch/step/log/evaluate/save/end
  - Confirmed `on_evaluate` receives metrics (after metrics-forwarding fix)

Refs: RHOAIENG-77626, RHOAIENG-54744

## Summary by CodeRabbit

* **New Features**
  * Added a unified Training Hub callback framework with a shared training context and standardized lifecycle hooks.
  * Introduced an adapter that routes Training Hub callback events (including metrics) into the Unsloth-backed training flow.
  * Extended `lora_sft`/LoRA SFT to accept custom callbacks and optional evaluation data and per-device eval batch sizing.
  * Re-exported callback components from the top-level package.
* **Bug Fixes**
  * Evaluation is now conditionally enabled only when evaluation data is provided.
* **Tests**
  * Added unit tests for callback behavior and adapter dispatch, including exception isolation and rank-based dispatch.
  * Added end-to-end callback smoke-check script and notebook.

### CODE REVIEW

**Comment by Sridhar1030** [comment]
@coderabbitai review

**Comment by RobotSail** [comment]
@training-hub-agent To your first point: this project only supports Python 3.11 or greater, never anything below Python 3.9. Are you sure this is necessary?

**Comment by Sridhar1030** [comment]
On exception isolation catching KeyboardInterrupt/SystemExit would break Ctrl-C / intentional exits, right? 

On empty logs  fixed (`logs is not None` + test).

On the eval params shallow copy  leaving as-is; `_prepare_dataset` doesn’t mutate nested params.

**Comment by efazal** [comment]
Thanks @Sridhar1030. The abstraction looks clean. Left a few comments to address.

File: tests/test_callbacks.py [hunk 1]
```diff
@@ -0,0 +1,290 @@
...
+        """TrainingHubCallback should be directly instantiable (not ABC)."""
+        cb = TrainingHubCallback()
+        assert cb is not None
+
+
+# ---------------------------------------------------------------------------
+# UnslothCallbackAdapter (requires transformers)
+# ---------------------------------------------------------------------------
+
+pytest.importorskip("transformers", reason="transformers not installed")
```
Comment (efazal) [inline tests/test_callbacks.py]: `pytest.importorskip` at module level causes all 6 `TestTrainingHubContext` and `TestTrainingHubCallback` tests to be skipped when `transformers` is not installed. Those tests have zero dependency on transformers. Move the `importorskip` call to just before `TestUnslothCallbackAdapter` so the base abstraction tests always run in lightweight CI.

File: src/training_hub/adapters/unsloth.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        logs: dict | None = None,
+    ) -> None:
+        """Dispatch to user callback with exception isolation and rank guard."""
+        if not state.is_world_process_zero:
+            return
+        try:
+            ctx = self._build_context(args, state, logs)
+            getattr(self._hub_callback, method_name)(ctx)
+        except Exception:
+            logger.exception(
```
Comment (efazal) [inline src/training_hub/adapters/unsloth.py:112]: When multiple callbacks are registered, this log doesn't say which one failed. Use `type(self._hub_callback).__name__` instead of the hardcoded `"TrainingHubCallback"` so the message reads e.g. `"SmokeLogger.on_step_end raised an exception"`.

File: src/training_hub/adapters/unsloth.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+    def _safe_call(
+        self,
+        method_name: str,
+        args: TrainingArguments,
+        state: TrainerState,
+        logs: dict | None = None,
+    ) -> None:
+        """Dispatch to user callback with exception isolation and rank guard."""
+        if not state.is_world_process_zero:
```
Comment (efazal) [inline src/training_hub/adapters/unsloth.py]: The rank-0 guard here is stricter than HF's default, which dispatches callbacks on all ranks and lets each callback decide. This means user callbacks can never run on worker ranks (e.g. per-node GPU memory logging). Fine for now since Unsloth LoRA is single-node, but consider adding a `run_on_all_ranks = False` class attribute on `TrainingHubCallback` as an opt-in before the FSDP/InstructLab adapters land.

File: src/training_hub/algorithms/lora.py [hunk 4]
```diff
@@ -543,8 +560,13 @@ def _build_training_args(self, params: Dict[str, Any],
             logging_steps=params.get('logging_steps', 1),
             logging_dir=params.get('tensorboard_log_dir'),
             save_steps=params.get('save_steps', 500),
+            eval_strategy="steps" if has_eval else "no",
             eval_steps=params.get('eval_steps', 500),
             save_total_limit=params.get('save_total_limit', 3),
+            per_device_eval_batch_size=params.get(
```
Comment (efazal) [inline src/training_hub/algorithms/lora.py:592]: nit: `per_device_eval_batch_size` is consumed here and listed in `get_optional_params`, but has no named parameter in `lora_sft()` or `LoRASFTAlgorithm.train()`. Add it as an explicit parameter alongside `eval_data_path` for discoverability.

File: src/training_hub/algorithms/lora.py [hunk 5]
```diff
@@ -707,6 +729,9 @@ def train(self,
               finetune_language_layers: Optional[bool] = None,
               # Model loading parameters
               trust_remote_code: Optional[bool] = None,
+              # Callback / eval parameters
+              callbacks: Optional[list[TrainingHubCallback]] = None,
```
Comment (efazal) [inline src/training_hub/algorithms/lora.py]: nit: passing a single callback (`callbacks=my_cb` instead of `[my_cb]`) produces a confusing `TypeError`. Normalize with `if isinstance(callbacks, TrainingHubCallback): callbacks = [callbacks]` early in the method.

File: src/training_hub/adapters/unsloth.py [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+        self._safe_call("on_train_end", args, state)
+
+
+def adapt_hub_callbacks(
+    callbacks: list[TrainingHubCallback],
+) -> list[TrainerCallback]:
+    """Convert a list of TrainingHubCallbacks to HuggingFace TrainerCallbacks.
+
+    Args:
+        callbacks: List of TrainingHubCallback instances.
```
Comment (efazal) [inline src/training_hub/adapters/unsloth.py:155]: nit: no type validation on list items. If a user passes a non-callback object, the error gets swallowed by exception isolation and the callback silently does nothing. Add an `isinstance` check in `__init__` with a clear `TypeError`.

File: src/training_hub/algorithms/lora.py [hunk 7]
```diff
@@ -543,8 +560,13 @@ def _build_training_args(self, params: Dict[str, Any],
             logging_steps=params.get('logging_steps', 1),
             logging_dir=params.get('tensorboard_log_dir'),
             save_steps=params.get('save_steps', 500),
+            eval_strategy="steps" if has_eval else "no",
```
Comment (efazal) [inline src/training_hub/algorithms/lora.py:589]: nit: when `eval_steps` is set but `eval_data_path` is not, `eval_steps` is silently ignored. A `logger.warning` here would save users debugging time.

File: tests/test_callbacks.py (hunk 1 above)
Comment (Sridhar1030) [reply tests/test_callbacks.py]: scoped to adapter tests via skipif

File: src/training_hub/adapters/unsloth.py (hunk 3 above)
Comment (Sridhar1030) [reply src/training_hub/adapters/unsloth.py]: added run_on_all_ranks

**Review by RobotSail — APPROVED** [review]
LGTM! Thanks for the excellent work on this @Sridhar1030 , and thank you @efazal for helping review!!
